File: export_tdx_all.py
#!/usr/bin/env python3
"""通达信所有日线数据导出到CSV

通达信日线数据格式 (.day 文件):
每条记录32字节:
- 日期 (4字节 int): 格式如 20250321
- 开盘价 (4字节 int): 实际值需除以100
- 最高价 (4字节 int): 实际值需除以100
- 最低价 (4字节 int): 实际值需除以100
- 收盘价 (4字节 int): 实际值需除以100
- 成交额 (4字节 float)
- 成交量 (4字节 int)
- 保留 (4字节 int)
"""

import struct
import os
import csv
from pathlib import Path
from datetime import datetime

# 通达信数据路径
TDX_VIPDOC_PATH = r"D:\new_tdx\vipdoc"

# 输出目录
OUTPUT_DIR = r"D:\新建文件夹\claude\tdx_output"

# CSV表头
CSV_HEADER = ['date', 'open', 'high', 'low', 'close', 'volume', 'amount']


def parse_day_file(filepath):
    """解析通达信 .day 日线文件"""
    data = []
    try:
        with open(filepath, 'rb') as f:
            while True:
                chunk = f.read(32)
                if len(chunk) < 32:
                    break
                # 解包: 7个数值 (i=int, f=float)
                vals = struct.unpack('iiiiiifi', chunk)

                date_val = vals[0]
                # 日期格式转换为 YYYY-MM-DD
                if date_val > 0:
                    date_str = str(date_val)
                    if len(date_str) == 8:
                        date_formatted = f"{date_str[:4]}-{date_str[4:6]}-{date_str[6:8]}"
                    else:
                        date_formatted = date_str
                else:
                    date_formatted = ""

                open_p = vals[1] / 100.0
                high = vals[2] / 100.0
                low = vals[3] / 100.0
                close = vals[4] / 100.0
                amount = vals[5]
                volume = vals[6]

                data.append([date_formatted, open_p, high, low, close, volume, amount])
    except Exception as e:
        print(f"  解析错误: {e}")
        return []
    return data


def to_csv(data, output_path, stock_code):
    """保存为CSV文件"""
    if not data:
        return False

    try:
        os.makedirs(os.path.dirname(output_path), exist_ok=True)
        with open(output_path, 'w', newline='', encoding='utf-8') as f:
            writer = csv.writer(f)
            writer.writerow(CSV_HEADER)
            writer.writerows(data)
        return True
    except Exception as e:
        print(f"  保存CSV失败: {e}")
        return False


def export_market(market, output_subdir, sample_dir=None):
    """导出单个市场的所有日线数据

    Args:
        market: 市场代码 ('sh' 或 'sz')
        output_subdir: 输出子目录名
        sample_dir: 样本输出目录 (保存前几个文件)
    """
    lday_path = os.path.join(TDX_VIPDOC_PATH, market, 'lday')

    if not os.path.exists(lday_path):
        print(f"路径不存在: {lday_path}")
        return 0

    day_files = [f for f in os.listdir(lday_path) if f.endswith('.day')]
    total = len(day_files)
    print(f"\n{market.upper()} 市场: 发现 {total} 个日线文件")

    success_count = 0
    error_count = 0
    sample_count = 0  # 样本文件计数

    for i, filename in enumerate(day_files, 1):
        filepath = os.path.join(lday_path, filename)
        stock_code = filename.replace('.day', '')

        # 解析数据
        data = parse_day_file(filepath)

        if data:
            # 输出到主目录
            csv_filename = f"{stock_code}.csv"
            output_path = os.path.join(OUTPUT_DIR, output_subdir, csv_filename)

            if to_csv(data, output_path, stock_code):
                success_count += 1

                # 保存前几个文件到样本目录
                if sample_dir and sample_count < 5:
                    sample_path = os.path.join(sample_dir, f"{market}_{csv_filename}")
                    to_csv(data, sample_path, stock_code)
                    sample_count += 1
            else:
                error_count += 1
        else:
            error_count += 1

        # 进度显示
        if i % 500 == 0 or i == total:
            print(f"  进度: {i}/{total} ({i*100//total}%)")

    print(f"  完成: 成功 {success_count}, 失败 {error_count}")
    return success_count


def main():
    print("=" * 60)
    print("       通达信日线数据导出工具")
    print("=" * 60)
    print(f"源路径: {TDX_VIPDOC_PATH}")
    print(f"输出目录: {OUTPUT_DIR}")

    # 样本输出目录
    sample_dir = r"D:\新建文件夹\claude\tdx-parser-workspace\iteration-1\eval-3\without_skill\outputs"
    os.makedirs(sample_dir, exist_ok=True)
    print(f"样本目录: {sample_dir}")

    # 确保输出目录存在
    os.makedirs(OUTPUT_DIR, exist_ok=True)

    # 统计开始时间
    start_time = datetime.now()

    # 导出上海市场
    sh_count = export_market('sh', 'sh', sample_dir)

    # 导出深圳市场
    sz_count = export_market('sz', 'sz', sample_dir)

    # 统计结束时间
    end_time = datetime.now()
    duration = (end_time - start_time).total_seconds()

    print("\n" + "=" * 60)
    print("导出完成!")
    print(f"上海市场: {sh_count} 个文件")
    print(f"深圳市场: {sz_count} 个文件")
    print(f"总计: {sh_count + sz_count} 个文件")
    print(f"耗时: {duration:.1f} 秒")
    print(f"输出目录: {OUTPUT_DIR}")
    print(f"样本文件: {sample_dir}")
    print("=" * 60)


if __name__ == '__main__':
    main()

File: export_tdx_all.js
const fs = require('fs');
const path = require('path');

// 通达信数据路径
const TDX_VIPDOC_PATH = 'D:/new_tdx/vipdoc';

// 输出目录
const OUTPUT_DIR = 'D:/新建文件夹/claude/tdx_output';

// 样本输出目录
const SAMPLE_DIR = 'D:/新建文件夹/claude/tdx-parser-workspace/iteration-1/eval-3/without_skill/outputs';

// CSV表头
const CSV_HEADER = 'date,open,high,low,close,volume,amount';

function readInt32LE(buffer, offset) {
    return buffer.readInt32LE(offset);
}

function readFloat(buffer, offset) {
    return buffer.readFloatLE(offset);
}

function formatDate(dateVal) {
    const dateStr = dateVal.toString();
    if (dateStr.length === 8) {
        return `${dateStr.slice(0,4)}-${dateStr.slice(4,6)}-${dateStr.slice(6,8)}`;
    }
    return dateStr;
}

function parseDayFile(filepath) {
    const buffer = fs.readFileSync(filepath);
    const records = [];
    const recordSize = 32;

    for (let i = 0; i + recordSize <= buffer.length; i += recordSize) {
        const date = readInt32LE(buffer, i);
        const open = readInt32LE(buffer, i + 4) / 100;
        const high = readInt32LE(buffer, i + 8) / 100;
        const low = readInt32LE(buffer, i + 12) / 100;
        const close = readInt32LE(buffer, i + 16) / 100;
        const amount = readFloat(buffer, i + 20);
        const volume = readInt32LE(buffer, i + 24);
        // const reserved = readInt32LE(buffer, i + 28);

        records.push({
            date: formatDate(date),
            open: open.toFixed(2),
            high: high.toFixed(2),
            low: low.toFixed(2),
            close: close.toFixed(2),
            amount: amount.toFixed(2),
            volume: volume
        });
    }

    return records;
}

function toCSV(data, filepath) {
    if (data.length === 0) return false;

    try {
        const rows = data.map(r =>
            `${r.date},${r.open},${r.high},${r.low},${r.close},${r.volume},${r.amount}`
        );
        const csv = [CSV_HEADER, ...rows].join('\n');

        fs.mkdirSync(path.dirname(filepath), { recursive: true });
        fs.writeFileSync(filepath, csv, 'utf8');
        return true;
    } catch (e) {
        console.error(`  保存失败: ${e.message}`);
        return false;
    }
}

function exportMarket(market, outputSubdir, sampleDir) {
    const ldayPath = path.join(TDX_VIPDOC_PATH, market, 'lday');

    if (!fs.existsSync(ldayPath)) {
        console.log(`路径不存在: ${ldayPath}`);
        return 0;
    }

    const files = fs.readdirSync(ldayPath).filter(f => f.endsWith('.day'));
    const total = files.length;
    console.log(`\n${market.toUpperCase()} 市场: 发现 ${total} 个日线文件`);

    let successCount = 0;
    let errorCount = 0;
    let sampleCount = 0;

    for (let i = 0; i < total; i++) {
        const file = files[i];
        const inputPath = path.join(ldayPath, file);
        const stockCode = file.replace('.day', '');

        try {
            const data = parseDayFile(inputPath);

            if (data.length > 0) {
                // 输出到主目录
                const outputFile = path.join(OUTPUT_DIR, outputSubdir, `${stockCode}.csv`);
                if (toCSV(data, outputFile)) {
                    successCount++;

                    // 保存前5个文件到样本目录
                    if (sampleDir && sampleCount < 5) {
                        const sampleFile = path.join(sampleDir, `${market}_${stockCode}.csv`);
                        toCSV(data, sampleFile);
                        sampleCount++;
                    }
                }
            } else {
                errorCount++;
            }
        } catch (e) {
            console.error(`  解析失败: ${file} - ${e.message}`);
            errorCount++;
        }

        // 进度显示
        if ((i + 1) % 500 === 0 || (i + 1) === total) {
            console.log(`  进度: ${i + 1}/${total} (${Math.floor((i + 1) * 100 / total)}%)`);
        }
    }

    console.log(`  完成: 成功 ${successCount}, 失败 ${errorCount}`);
    return successCount;
}

function main() {
    console.log('='.repeat(60));
    console.log('       通达信日线数据导出工具');
    console.log('='.repeat(60));
    console.log(`源路径: ${TDX_VIPDOC_PATH}`);
    console.log(`输出目录: ${OUTPUT_DIR}`);
    console.log(`样本目录: ${SAMPLE_DIR}`);

    const startTime = Date.now();

    // 确保输出目录存在
    fs.mkdirSync(OUTPUT_DIR, { recursive: true });
    fs.mkdirSync(SAMPLE_DIR, { recursive: true });

    // 导出上海市场
    const shCount = exportMarket('sh', 'sh', SAMPLE_DIR);

    // 导出深圳市场
    const szCount = exportMarket('sz', 'sz', SAMPLE_DIR);

    // 尝试导出其他市场（如果存在）
    const bjCount = exportMarket('bj', 'bj', null);
    const dsCount = exportMarket('ds', 'ds', null);

    const endTime = Date.now();
    const duration = ((endTime - startTime) / 1000).toFixed(1);

    console.log('\n' + '='.repeat(60));
    console.log('导出完成!');
    console.log(`上海市场: ${shCount} 个文件`);
    console.log(`深圳市场: ${szCount} 个文件`);
    console.log(`北京市场: ${bjCount} 个文件`);
    console.log(`大宗交易: ${dsCount} 个文件`);
    console.log(`总计: ${shCount + szCount + bjCount + dsCount} 个文件`);
    console.log(`耗时: ${duration} 秒`);
    console.log(`输出目录: ${OUTPUT_DIR}`);
    console.log(`样本文件: ${SAMPLE_DIR}`);
    console.log('='.repeat(60));
}

main();
